lib/data_freshness/main.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
+
+import asyncio
+import os
+from datetime import datetime, timedelta
+from pngme.api import AsyncClient
+
+
+async def get_data_freshness(
+    api_client: AsyncClient,
+    user_uuid: str,
+    utc_starttime: datetime,
+    utc_endtime: datetime,
+) -> float:
+    """Return the time in minutes between utc_endtime and the most recent financial event or alert,
+    as an indicator of data freshness.
+    Args:
+        api_client: Pngme Async API client
+        user_uuid: Pngme mobile phone user_uuid
+        utc_starttime: the time from which balances are considered
+        utc_endtime: the time until which balances are considered
+        page: the page(s) requested
+    Returns:
+        Return the time in minutes between utc_endtime and the most recent financial event or alert
+    """
+
+    # STEP 1: fetch list of institutions belonging to the user
+    institutions = await api_client.institutions.get(user_uuid=user_uuid)
+
+    # STEP 2: get a list of all transactions for each institution
+    transaction_inst_coroutine = []
+    balance_inst_coroutines = []
+    alerts_inst_coroutines = []
+    for inst in institutions:
+        transaction_inst_coroutine.append(
+            api_client.transactions.get(
+                user_uuid=user_uuid,
+                institution_id=inst.institution_id,
+                utc_starttime=utc_starttime,
+                utc_endtime=utc_endtime,
+                page=1,
+            )
+        )
+        balance_inst_coroutines.append(
+            api_client.balances.get(
+                user_uuid=user_uuid,
+                institution_id=inst.institution_id,
+                utc_starttime=utc_starttime,
+                utc_endtime=utc_endtime,
+                page=1,
+            )
+        )
+        alerts_inst_coroutines.append(
+            api_client.balances.get(
+                user_uuid=user_uuid,
+                institution_id=inst.institution_id,
+                utc_starttime=utc_starttime,
+                utc_endtime=utc_endtime,
+                page=1,
+            )
+        )
+
+    transactions_per_institution = await asyncio.gather(*transaction_inst_coroutine)
+    balances_per_institution = await asyncio.gather(*balance_inst_coroutines)
+    alerts_per_institution = await asyncio.gather(*alerts_inst_coroutines)
+
+
+    # STEP 3: flatten all balances, transactions, and alerts from all institutions, and get most recent time
+    transactions_flattened = []
+    for ix, transactions in enumerate(transactions_per_institution):
+        institution_id = institutions[ix].institution_id
+        for transaction in transactions:
+            transactions_flattened.append(
+                dict(transaction, institution_id=institution_id))
```

Comment:
Any reason each is casted to `dict` instead of left as a typed object?

Reply:
Following convention from other features, but if there is a simpler approach I'd love to see it!